### ISSUE TITLE
New windows from RubyMine are not drawn

### ISSUE THREAD

**Issue by danuker** [i3/i3#1784]
1. Open a project with RubyMine (or probably any other JetBrains IDE), and switch i3 to either **tabbed** or **stacked** mode.

2. Select two files from the Project panel (left)
![diff-p1](https://cloud.githubusercontent.com/assets/74835/8523039/52c5ab7a-23f3-11e5-9553-4ec55c613e55.png)

3. Press Ctrl+D (for the difference between the files). The new window title **appears and is focused**, but the window itself is **not drawn** (the old one remains).
![diff-p2](https://cloud.githubusercontent.com/assets/74835/8523038/52c3821e-23f3-11e5-9695-eb5c9afbe224.png)

4. Just moving the mouse on top of the new tab (not even clicking) makes the window re-draw and show up properly.
![diff-p3](https://cloud.githubusercontent.com/assets/74835/8523040/52c6b8c6-23f3-11e5-9f41-632d99fb19a8.png)

Step 4 should not be necessary - the window should show up by itself.

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow http://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by i3bot** [comment]
I don’t see a version number. Could you please copy & paste the output of `i3 --version` into this issue?

**Comment by danuker** [comment]
i3 version 4.10.2 (2015-04-16, branch "4.10.2") © 2009-2014 Michael Stapelberg and contributors

I will provide a log file later today if needed.

**Comment by Airblader** [comment]
Yes, please provide a log file. Your version doesn't include the hidden state handling we will introduce yet, so that can't be it. Since i3 isn't involved in the client rendering, this could be an application bug. The only thing i3 could be doing wrong is not correctly mapping the clients.

**Comment by Airblader** [comment]
Please make sure to read [the debugging guide](https://github.com/i3/i3/blob/next/docs/debugging) from the current dev branch (not the website) to make it a bit easier on us.

**Comment by danuker** [comment]
http://logs.i3wm.org/logs/5739238230327296.bz2

At around 16:00:00 I opened "file1 vs file2 (/home/danuth/RubymineProjects/untitled)" in the default mode (and the window was drawn).
At around 16:00:10 I switched to tabbed mode, and opened it again. (and the window was not drawn).
At around 16:00:20 I moved my mouse up on the tab, and the window was rendered.

**Comment by Airblader** [comment]
Can you please also do something like 

sleep 10 && i3-msg -t get_tree > /tmp/i3-tree-before.json && sleep 5 && i3-msg -t get_tree > /tmp/i3-tree-after.json

in the background, open the files in tabbed mode and wait until the first i3-msg command is finished (without getting them to draw correctly), then move the mouse over the tab and wait for the second i3-msg command to finish. Finally, upload those files or paste them here. Make sure to have nothing else open anywhere to keep the files small and your data confidential.

Thanks!

**Comment by danuker** [comment]
Before: http://pastebin.com/JXSR6CVs
After: the same (md5sum shows same hash).

**Comment by danuker** [comment]
When I move the mouse to another window instead of the focused tab, the only difference is the focused window.

**Comment by Airblader** [comment]
So i3 at least thinks the correct window is focused.

Can you run `sleep 10 && xwininfo -all -id $(xdotool getactivewindow) > /tmp/wininfo.log` so it writes the file during the broken state? This should tell us whether the (correct) window is mapped.

**Comment by danuker** [comment]
It seems the original window is active (not the new Diff).

http://pastebin.com/KE2T2NBF

**Comment by danuker** [comment]
When I press keys in the broken state, the original window's line:column indicators update.
But they're crazy and nowhere near where I left it before opening the new one.

**Comment by danuker** [comment]
The original window shows my cursor's position in the new (invisible) diff window.

**Comment by Airblader** [comment]
Thanks, your cooperation is much appreciated!

One more thing, can you upload your i3 config (or alternatively see if the same issue arises with the default config)?

**Comment by Airblader** [comment]
Also looking at the output this could have something to do with something we fixed recently. On mobile I can't check if the release contained that fix already, but maybe you could compile the current development branch and see if you can still reproduce the issue there? Not that we end up chasing ghosts here.

**Comment by danuker** [comment]
The same happens with the default config.
I'll check the latest branch at home (in about two hours).

Glad to help! Thank you for your time also :+1:

**Comment by danuker** [comment]
My branch is up-to-date with 'origin/next'.

make. sudo make install

Still have bug. 
http://logs.i3wm.org/logs/5764017373052928.bz2

Happens around 
_NET_WM_NAME changed to "Gemfile vs Gemfile.lock (/home/dan/PycharmProjects/RubyHDR)

**Comment by stapelberg** [comment]
> My branch is up-to-date with 'origin/next'.

Please provide the output of “i3 --moreversion” instead, that is much more precise (as it does not depend on the current time/contents of a branch).

**Comment by danuker** [comment]
i3 --moreversion
Binary i3 version:  4.10.2-223-gb30d87e (2015-07-01, branch "next") © 2009 Michael Stapelberg and contributors
Running i3 version: 4.10.2-223-gb30d87e (2015-07-01, branch "next") (pid 13114)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3-with-/i3

**Comment by danuker** [comment]
I'm going to sleep guys (don't wait up). Back in ~8h

**Comment by Airblader** [comment]
Things don't always move this fast anyway. I won't have time for a few days due to work. But maybe someone else will take a look. :)

**Comment by Airblader** [comment]
I just wanted to see if I can reproduce this with Webstorm. It's interesting – Pressing Ctrl+D on two files in tabbed layout, I do see the new window open, but clicking anywhere on the new window immediately focuses the first tab. If I manually select the second tab again, it doesn't do that anymore.

I'm going to go on the assumption that it's the same issue, though, so this might make it easier to work on it.

**Comment by danuker** [comment]
Yes, that happens to me as well - when clicking on the window's space, the original (parent) window title gets focused again. I also think it's the same bug.

I'm glad there isn't a crash and that it recovers somewhat gracefully :)

Maybe it has to do with the way Java handles stuff - the IntelliJ stuff is written using Swing.
Source code for IntelliJ (their Java IDE) is available here:
https://github.com/JetBrains/intellij-community

Thanks for taking the time to reproduce it! I know how valuable time is, especially when working.

**Comment by catchin** [comment]
I have the same problem and can reproduce it with version: 4.11 (2015-09-30, branch "4.11") (4.11-1 on debian testing).
The "i3-msg -t get_tree" command results in the same file before and after the window is in foreground.

Here is my debug log file:
http://logs.i3wm.org/logs/5673385510043648.bz2
At "17:56:12" I opened the diff window and at "17:56:18" I moved the mouse to the title bar and it popped into the foreground.

Any news on this issue?

**Comment by TomRauchenwald** [comment]
I encountered the same issue with IntelliJ IDEA.
As a data point: It doesn't happen with 1.8 (i3 version 4.8 (2014-06-15, branch "4.8") © 2009-2014 Michael Stapelberg and contributors), but I could reproduce it with 1.9.

**Comment by TomRauchenwald** [comment]
I bisected this now.

8a618e4b006000cbd521031f4c191e862da03914 is the first bad commit
commit 8a618e4b006000cbd521031f4c191e862da03914
Author: Tony Crisci <tony@dubstepdish.com>
Date:   Tue Jun 17 02:49:39 2014 -0400

    bugfix: don't set input focus if not accepted
    
    http://tronche.com/gui/x/icccm/sec-4.html#s-4.1.7
    
    > Clients using the Globally Active model can only use a SetInputFocus request
    > to acquire the input focus when they do not already have it on receipt of one
    > of the following events:
    > * ButtonPress
    > * ButtonRelease
    > * Passive-grabbed KeyPress
    > * Passive-grabbed KeyRelease
    
    Since managing a window happens on a MapNotify (which is absent from this
    list), the window cannot accept input focus, so we should not try to focus
    the window at all.
    
    Fixes an issue with xfce4-notifyd which (correctly) declines focus when
    we send WM_TAKE_FOCUS, which puts i3 in a state where i3 focus and X
    focus are different when a notification appears.

**Comment by Airblader** [comment]
Thanks for the bisect, that's very useful!

**Comment by LemonBoy** [comment]
I'm experiencing this problem, or a very similar one, too with i3 (but the same happens with dwm too).
If I open a wireshark window it doesn't get redrawn at regular intervals (eg, the blinking bar in the text fields stays still for a handful of seconds, then disappears to reappear after some time). Switching focus continuously makes the window redraw regularly-ish.
The output of i3 -v is `i3 version 4.11 (2015-09-30, branch "4.11") © 2009 Michael Stapelberg and contributors`.

**Comment by leonardinius** [comment]
I've also encountered the same behavior with IntelliJ IDEA. 
```
i3 --version
i3 version 4.10.3 (2015-07-30, branch "4.10.3") © 2009-2014 Michael Stapelberg and contributors
```

**Comment by leonardinius** [comment]
Quick question though: is there any workaround I could use or something?..

**Comment by Airblader** [comment]
The workaround would be to locallly revert the commit mentioned above.

**Comment by leonardinius** [comment]
@Airblader It actually worked. Thanks.

**Comment by Airblader** [comment]
@leonardinius Sorry for the late reply, I missed your question.

> Is the proper bugfix planned in some future release?

We'd love to fix the bug if it can be fixed reasonably in i3. I'm not yet sure what the fix looks like since the bad commit tries to make i3 follow the spec better and as far as I can tell, everything about it sounds correct to me. So I'm not sure i3 is really at fault here or what exactly is going wrong.

As always, contributions are welcome.

**Comment by AMPivovarov** [comment]
Hi,

I don't know much about Xorg, but it looks like in this case the [spec](http://www.sbin.org/doc/Xlib/chapt_12.html) had been read in a wrong way.

"Globally Active Input" should not be mixed with "No Input" model.

While the first one indeed means "window should not take focus", the second one means "window should decide whether it want take focus". (`A clock, for example, or a text editor with no currently open frames might not want to take focus even though the window manager generally believes that clients should take the keyboard focus`)
So instead of marking container focused (and making an attempt to forcefully set focus in X11 later), i3 should send WM_TAKE_FOCUS signal to the window and listen for the reply via 'XSetInputFocus' (that could happen on one of the children windows).

`Clients using the Globally Active model can only use XSetInputFocus() to acquire the input focus when they do not already have it on receipt of one of the following events: ButtonPress, ButtonRelease, passive-grabbed KeyPress, and passive-grabbed KeyRelease.` - means "Clients should not steal focus by themself, except as reaction at <...> events and only if they really need to (`In general, clients should avoid using passive-grabbed key events for this purpose except when they are unavoidable`)

I'm not sure if i get the idea right, but the model table might look like this:

                    |  Input - True  |  Input - False
--------------------|----------------|-----------------
WM_TAKE_FOCUS - ON | Locally Active | Globally Active
WM_TAKE_FOCUS - OFF | Passive | No Input

**Comment by Airblader** [comment]
Your interpretation sounds good to me, I think that might be it. Can you please try and see if the following patch fixes the issue?

````
diff --git a/src/manage.c b/src/manage.c
index 05ac15f..5fe9368 100644
--- a/src/manage.c
+++ b/src/manage.c
@@ -557,6 +557,10 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
         con_focus(nc);
     }
 
+    if (nc->window->doesnt_accept_focus && nc->window->needs_take_focus) {
+        send_take_focus(nc->window->id, last_timestamp);
+    }
+
     tree_render();
 
     /* Windows might get managed with the urgency hint already set (Pidgin is

```

**Comment by AMPivovarov** [comment]
@Airblader No luck.
And this case is already checked in x.c@1101 (x_push_changes).

Wouldn't it be a better way to fix 8a618e4?
```
diff --git a/src/manage.c b/src/manage.c
index 05ac15f..993e4a8 100644
--- a/src/manage.c
+++ b/src/manage.c
@@ -550,9 +550,13 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
         xcb_discard_reply(conn, wm_user_time_cookie.sequence);
     }
 
+    if (set_focus && nc->window->doesnt_accept_focus && !nc->window->needs_take_focus) {
+        set_focus = false;
+    }
+
     /* Defer setting focus after the 'new' event has been sent to ensure the
      * proper window event sequence. */
-    if (set_focus && !nc->window->doesnt_accept_focus && nc->mapped) {
+    if (set_focus && nc->mapped) {
         DLOG("Now setting focus.\n");
         con_focus(nc);
     }
```
<br/>
Another issue: even after reverting 8a618e4, closing the window causes the main one to gain focus "partially" (only a couple of shortcuts works).
This might be Intellij issue - [IDEA-112607](https://youtrack.jetbrains.com/issue/IDEA-112607), [IDEA-147626](https://youtrack.jetbrains.com/issue/IDEA-147626), [IDEA-65637](https://youtrack.jetbrains.com/issue/IDEA-65637), but in i3 4.7 it almost never happen.

**Comment by Airblader** [comment]
Yes, I think it looks like we currently don't handle the case of rejected WM_TAKE_FOCUS correctly.

**Comment by Airblader** [comment]
So this is a bit of a long shot but this fixes the issue:

````
diff --git a/src/manage.c b/src/manage.c
index f868c85..fc45763 100644
--- a/src/manage.c
+++ b/src/manage.c
@@ -604,7 +604,7 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
 
     /* Defer setting focus after the 'new' event has been sent to ensure the
      * proper window event sequence. */
-    if (set_focus && !nc->window->doesnt_accept_focus && nc->mapped) {
+    if (set_focus && (!nc->window->doesnt_accept_focus || nc->window->needs_take_focus) && nc->mapped) {
         DLOG("Now setting focus.\n");
         con_focus(nc);
     }
diff --git a/src/x.c b/src/x.c
index 0fbc4d1..1aedd7b 100644
--- a/src/x.c
+++ b/src/x.c
@@ -1124,7 +1124,7 @@ void x_push_changes(Con *con) {
                      to_focus, focused, focused->name);
                 send_take_focus(to_focus, last_timestamp);
 
-                ewmh_update_active_window((con_has_managed_window(focused) ? focused->window->id : XCB_WINDOW_NONE));
+                ewmh_update_active_window(XCB_WINDOW_NONE);
 
                 if (to_focus != last_focused && is_con_attached(focused))
                     ipc_send_window_event("focus", focused);
```

A coworker of mine is going to be trying this patch for unwanted side effects.

**Comment by catchin** [comment]
I just tried it with the latest IntelliJ IDEA and it fixes the problem that the new window is not drawn. So this works :).

However, when closing the new window, the main window is not fully focused. Meaning that it is shown and i3 title bar indicates it should be focused, but it does not respond to any key press.

**Comment by klassm** [comment]
I didn't notice that behaviour today - I'll try to confirm that tomorrow though.

**Comment by Airblader** [comment]
@catchin Thanks for the feedback. I can reproduce this, but oddly it only starts happening the second time I open the diff window. The first time I open and close it, the focus is set correctly.

**Comment by Airblader** [comment]
I think when handling FocusIn events, we now need to update `_NET_ACTIVE_WINDOW`. I don't think we need to update the input focus since that's what should've caused it in the first place. Anyway, this is just a reference for a potential proper fix.

@catchin I can see the behavior you describe even prior to this patch. Furthermore, I've looked into this for a while now and I believe that this is not an i3 issue. After sending `WM_TAKE_FOCUS`, we expect a `FocusIn` (with certain properties) caused by the client's `XSetInputFocus` request. We do receive the event, but i3 ignores it – for, AFAICT, valid reasons.

I then exchanged the conditions causing it to be ignored with the conditions awesomeWM uses. Same effect.

Can somebody try whether RubyMine (or IntelliJ) even properly takes focus in this scenario when using *other* window managers?

**Comment by Airblader** [comment]
I've looked into how awesome and Openbox handle this and they handle all of this quite a bit different. I'll see whether things improve if we do it the same way.

**Comment by Airblader** [comment]
OK. Before I go to bed now, here is where I ended up for now: In general, the way i3 handles the input hint and `WM_TAKE_FOCUS` sounds reasonable. There's only two points:

* We probably shouldn't update `_NET_ACTIVE_WINDOW` if we are only just sending `WM_TAKE_FOCUS` since it may be rejected. This is really mostly cosmetics, though, but should be fixed.
* I don't like that i3 suppresses `WM_TAKE_FOCUS` if the client accepts input because "it's unnecessary in that case". Neither Openbox nor awesome do this. I think a client should either set the hints accordingly or deal with the request it gets.

But as far as the actual issue here goes, it is, for all I can tell, in fact 8a618e4b006000cbd521031f4c191e862da03914 which is wrong. Generally even if the client doesn't accept input, we still want to at least focus the container. `src/x.c` should and will take care of not actually setting the input focus. However, I get the idea @acrisci was following here, and in fact KDE and Openbox do this too – but it should only apply to the globally active model. My patch above (the first part of it) fixes this and I now believe that it fixes it properly.

That just leaves us with focus not being "restored" correctly. Here's a couple points:
* I verified that we sent `WM_TAKE_FOCUS` to the client. So they should be calling `XSetInputFocus`.
* `XSetInputFocus` bypasses the window manager. We aren't even involved!
* … except that it triggers a few events, like `FocusIn`. I verified that we end up ignoring this event.
So what we should look at here is whether we set input focus on anything, overriding the input focus RubyMine acquired for itself.

**Comment by Airblader** [comment]
I am splitting this issue up into the separate issues we identified. I also will not change that we suppress `WM_TAKE_FOCUS` if we are allowed to set input focus ourselves for the time being because it doesn't violate the specification and was part of a bugfix.

I will provide a PR that fixes the IntelliJ / RubyMine issue properly, but before I do I want to look into the issue when closing the diff window again. I think we simply update the focus, effectively overriding what IntelliJ did (and that might not be trivial to fix), but while also orthogonal to the issue here I want to at least understand what's happening first.

**Comment by Airblader** [comment]
OK, I've looked more into this now. Using `$mod+q` as my binding to close the diff view, I noticed that focus is restored correctly if I press `$mod` after the window is closed (or keep it held down for long enough).

I then ran RubyMine under xtrace. When focus is "lost", it puts out a single `SetInputFocus` request with a set timestamp, so this is probably it responding to our `WM_TAKE_FOCUS`. However, this request is denied and RubyMine just doesn't care. When focus isn't lost, we see the same request, but followed another (or two) `SetInputFocus` requests for CurrentTime, which then are successful.

In other words I believe this to be a RubyMine bug now because the request is ignored by the server and RubyMine doesn't bother to check that. But since it doesn't allow us to set input focus, we respect that and don't do it.

**Comment by AMPivovarov** [comment]
>However, this request is denied and RubyMine just doesn't care.
What should it do after its request was denied? Why are they denied, btw?

>RubyMine shouldn't set WM_HINTS.input = 0
This flag is set by JDK.

**Comment by Airblader** [comment]
> What should it do after its request was denied? Why are they denied, btw?

I haven't looked at that, mostly because it's close to impossible. All I can see is that they send the request and that i3 doesn't do anything to change the input focus. I can see that there is some unmapping and window destroying going on at the same time, so my best guess is that RubyMine is bringing things out of order. But it'd be easier of the IntelliJ guys debugged this in detail.

**Comment by Airblader** [comment]
Cancel that! I think I found the bug for this in i3 now. :-)

**Comment by Airblader** [comment]
Well, sorry for jumping the gun, but I was wrong. I thought it might be an issue that we send a "last timestamp" in `WM_TAKE_FOCUS` instead of CurrentTime, but that is actually exactly how we *have* to do it according to the spec.

Still… maybe we are sending a "too old" timestamp because we forgot to update our last saved timestamp? If we had a focus change since then, this would cause RubyMine's request to be denied.

I'll look into that.

**Comment by AMPivovarov** [comment]
Just merged your branch.
Note, that the focus is not lost completely, when diff window is closed. 
Global actions (like "Find Action" or "Go To Class") can be invoked and (If the focus should be in editor) caret can be moved (but it is invisible and typing doesn't work).
The issue is very similar to the https://youtrack.jetbrains.com/issue/IDEA-65637 and https://youtrack.jetbrains.com/issue/IDEA-147148, so it might be an intellij bug indeed (AFAIU, WM can't break focus partially).

**Comment by acrisci** [comment]
#2315

**Comment by Airblader** [comment]
@AMPivovarov I've now verified that the timestamps aren't the issue either. So I maintain my previous position that this is probably an IntelliJ issue. I've said it before, but this focus stuff is quite confusing…

@acrisci I'm not sure why you link that again, but as far as the current discussion goes, this is unrelated. IntelliJ *does* send the request, so they don't reject `WM_TAKE_FOCUS` here. For some reason they just fail to acquire the input focus from the X server.

Anywho. This issue will be fixed with #2317. #2315 is a separate issue, as is the focus restoring when closing the diff.

**Comment by acrisci** [comment]
> I don't like that i3 suppresses WM_TAKE_FOCUS if the client accepts input because "it's unnecessary in that case". Neither Openbox nor awesome do this. I think a client should either set the hints accordingly or deal with the request it gets.

See #1167 and 05f0585

**Comment by Airblader** [comment]
@acrisci Yeah, I've stated later on that I don't intend to change this because of that bugfix.

**Comment by AMPivovarov** [comment]
@Airblader 
Another observation: closed window not necessary should belong to the IDE.
Closing terminal window, that results in IDE window being focused, leads to the same issue. But regular i3 actions "focus left/right" / "focus on mouse over" work fine.
Is there are any difference between these cases in i3 focus handling?

**Comment by Airblader** [comment]
@AMPivovarov No, there's no difference. There's only very little information that connects different windows on some level, but that information is not relevant for the focus handling at all.

**Comment by leonardinius** [comment]
Just to confirm:
- version `4.12` from ubuntu 16.04 - does not work as expected
- local buld from `next` 012de8c - work as expected (y)
